### PR TITLE
Refactor server struct

### PR DESCRIPTION
**Pull request by bkielbasa**
The idea behind it is to add unit tests to server and to refactor it to make it independent from outside services like ZooKeeper

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -16,27 +16,37 @@ import (
...
 	storage  Storager
-	le       *leaderElection
+	le       LeaderElectioner
+	ctx context.Context
+	cancelCtx context.CancelFunc
 }
 
-func NewServer(host string, port int, zkhosts []string) (*Server, error) {
+type LeaderElectioner interface {
+	Close()
```
[the remark this answers is not in the document]
Comment (bkielbasa) [reply server/server.go]: yeah, for now. I'll add `IsMaster()` and `IsSlave()` functions. I do not need them now so I did not add them yet.

File: server/server.go [hunk 2]
```diff
@@ -105,40 +127,50 @@ func (s *Server) handleClientConnection(conn net.Conn) {
 	reader := bufio.NewReader(conn)
 	tp := textproto.NewReader(reader)
 
+	ct, cancel := context.WithCancel(s.ctx)
+
 	for {
```
[the remark this answers is not in the document]
Comment (bkielbasa) [reply server/server.go]: yeah, or I'll use other design patterns. It's definitely requires refactoring but I have no good idea to do that now.

**Comment by bkielbasa** [comment]
As action point, I created an issue with refactoring it https://github.com/genki-dama/genki/issues/10

File: server/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmierzwa) [reply server/server.go]: top-down decomposition could be good for a start (starting from each case)
as a rule of thumb, I try having a single line of code (preferably a single function call), in each case statement
then, maybe some patterns will be more visible from the extracted functions (i.e. common parameters or parent class fields that suggest extracting a new aggregating type)

**Pull request merged**